File: static/xor/xor.pl
#!/usr/bin/perl
use strict;
use POSIX;
use Math::BigInt;

my $code="Cryptkey";
my $ctxt="a7a02122c7ad1477c7e1b0b91ea46fbbb03054b87a00f0eead";

my $ptext = decrypt($code,$ctxt);
printf("text=%s\n",$ptext);
my $ctext =encrypt($code,$ptext);
printf("text=%s\n",$ctext);
printf("text=%s\n",decrypt($code,$ctext));

sub encrypt
{
   my $code=shift;
   my $text=shift;
   my $prand="";
   foreach my $c (split(//,$code)){ $prand.=ord($c);}
   my $sPos=int(length($prand)/5);
   my $mult=int(substr($prand,$sPos,1).
                substr($prand,$sPos*2,1).
                substr($prand,$sPos*3,1).
                substr($prand,$sPos*4,1).
                substr($prand,$sPos*5,1));
   my $incr=ceil(length($code)/2);
   my $modu=2**31-1;
   my $salt=sprintf("%.0f",(rand()*1000000000)%100000000);
   $prand.=$salt;
   while(length($prand)>10){
      my $a=Math::BigInt->new(substr($prand,0,10));
      my $b=Math::BigInt->new(substr($prand,10));
      $prand=($a+$b);
   }
   $prand=sprintf("%d",$prand);
   $prand=($mult*$prand+$incr)%$modu;
   my $enc_str="";
   foreach my $c (split(//,$text)){
      my $x = floor(($prand / $modu) * 255);
      my $enc_chr = int(ord($c)^$x);
      $enc_str .= sprintf("%02x", $enc_chr);
      $prand=($mult*$prand+$incr)%$modu;
   }
   $salt = sprintf("%08x",$salt);
   $enc_str.=$salt;
   return($enc_str);
}

sub decrypt
{
   my $code=shift;
   my $ctext=shift;
   my $prand="";
   foreach my $c (split(//,$code)){ $prand.=ord($c);}
   my $sPos=int(length($prand)/5);
   my $mult=int(substr($prand,$sPos,1).
                substr($prand,$sPos*2,1).
                substr($prand,$sPos*3,1).
                substr($prand,$sPos*4,1).
                substr($prand,$sPos*5,1));
   my $incr=ceil(length($code)/2);
   my $modu=2**31-1;
   my $salt=hex(substr($ctext,length($ctext)-8));
   my $str=substr($ctext,0,length($ctext)-8);
   $prand.=$salt;
   while(length($prand)>10){
      my $a=Math::BigInt->new(substr($prand,0,10));
      my $b=Math::BigInt->new(substr($prand,10));
      $prand=($a+$b);
   }
   $prand=sprintf("%d",$prand);
   $prand=($mult*$prand+$incr)%$modu;
   my $enc_str="";
   for(my $i=0;$i<length($str);$i+=2){
      my $h=substr($str,$i,2);
      my $d=hex($h);
      my $x = floor(($prand / $modu) * 255);
      $enc_str.=chr($d ^ $x);
      $prand=($mult*$prand+$incr)%$modu;
   }
   return($enc_str);
}

1;

File: static/xor/xor.js
function encrypt(str, pwd){
   var prand = "";
   for (var i=0;i<pwd.length;i++){
      prand+=pwd.charCodeAt(i).toString();
   }
   var sPos=Math.floor(prand.length / 5);

   var mult=parseInt(prand.charAt(sPos)+prand.charAt(sPos*2)+
            prand.charAt(sPos*3)+prand.charAt(sPos*4)+
            prand.charAt(sPos*5));
   if (mult<2){
      alert("Algorithm cannot find a suitable hash. "+
            "Please choose a different password. \n"+
            "Possible considerations are to choose "+
            "a more complex or longer password.");
      return null;
   }

   var incr=Math.ceil(pwd.length / 2);
   var modu=Math.pow(2, 31) - 1;
   var salt=Math.round(Math.random()*1000000000)%100000000;
   prand+=salt;
   while(prand.length>10){
      var a=parseBigInt(prand.substring(0, 10));	   
      var b=parseBigInt(prand.substring(10, prand.length));	   
      var c=a.add(b);
      prand=c.toString();
   }
   prand=(mult*prand+incr)%modu;
   var enc_chr="";
   var enc_str="";
   for (var i=0;i<str.length;i++){
      enc_chr=parseInt(str.charCodeAt(i)^Math.floor((prand / modu) * 255));
      if (enc_chr<16){
         enc_str+="0"+enc_chr.toString(16);
      } 
      else{
         enc_str += enc_chr.toString(16);
      }
      prand = (mult * prand + incr) % modu;
   }
   salt = salt.toString(16);
   while(salt.length<8){
      salt = "0" + salt;
   }
   enc_str += salt;
   return(enc_str);
}

function decrypt(str, pwd) {
   if (str==null || str.length<8){
      alert("A salt value could not be extracted from the encrypted "+
            "message because it's length is too short. "+
            "The message cannot be decrypted.");
      return;
   }
   if (pwd==null || pwd.length<=0){
      alert("Please enter a password with which to decrypt the message.");
      return;
   }
   var prand = "";
   for(var i=0;i<pwd.length;i++){
      prand += pwd.charCodeAt(i).toString();
   }
   var sPos=Math.floor(prand.length / 5);
   var mult=parseInt(prand.charAt(sPos)+
            prand.charAt(sPos*2)+prand.charAt(sPos*3)+
            prand.charAt(sPos*4) + prand.charAt(sPos*5));
   var incr=Math.round(pwd.length / 2);
   var modu=Math.pow(2, 31) - 1;
   var salt=parseInt(str.substring(str.length - 8, str.length), 16);
   str=str.substring(0, str.length - 8);
   prand+=salt;
   while(prand.length>10){
      var a=parseBigInt(prand.substring(0, 10));	   
      var b=parseBigInt(prand.substring(10, prand.length));	   
      var c=a.add(b);
      prand=c.toString();
   }
   prand=(mult*prand+incr)%modu;
   var enc_chr="";
   var enc_str="";
   for(var i=0;i<str.length;i+=2){
      enc_chr=parseInt(parseInt(str.substring(i,i+2),16)^
              Math.floor((prand/modu)* 255));
      enc_str+=String.fromCharCode(enc_chr);
      prand = (mult * prand + incr) % modu;
   }
   return(enc_str);
}

function parseBigInt(str) {
   return new BigInteger(str,16);
}

function parseBigInt(str,r) {
   return new BigInteger(str,r);
}

